File: compiler/bench/collatz.fs
\ expected: 1637505
\ Collatz sequence stopping time sum benchmark

: collatz-len ( n -- steps )
  0 swap
  begin dup 1 > while
    dup 2 mod 0= if
      2/
    else
      3 * 1+
    then
    swap 1+ swap
  repeat drop ;

: main 0 100001 1 do i collatz-len + loop . cr ;

File: compiler/bench/collatz.c
// expected: 1637505
#include <stdio.h>
long collatz_len(long n) {
    long steps = 0;
    while (n > 1) {
        if (n % 2 == 0) n /= 2;
        else n = 3 * n + 1;
        steps++;
    }
    return steps;
}
int main() {
    long sum = 0;
    for (long i = 1; i < 100001; i++)
        sum += collatz_len(i);
    printf("%ld\n", sum);
    return 0;
}
